### PR TITLE
Add `COOKIE_SAME_SITE` option

### PR DESCRIPTION
**Pull request by yhwang**
Add an option to control the SameSite attribute of the
`Set-Cookie` response header. Default value is `Strict`.

### CODE REVIEW

**Comment by yhwang** [comment]
FYI @yanniszark I created this PR and then I saw yours.

**Comment by animeshsingh** [comment]
@yanniszark can we have someone prioritize this?

**Comment by yhwang** [comment]
@yanniszark it would be good to make it configurable. I am using `Strict` as default value, if you prefer `Lax`, I can change it.

File: README.md [hunk 1]
```diff
@@ -96,6 +96,7 @@ Session store-related settings:
 | - | - | - |
 | `SESSION_STORE_PATH` | "/var/lib/authservice/data.db" | Path to local session store. Backed by BoltDB. |
 | `SESSION_MAX_AGE` | "86400" | Time in seconds after which sessions expire. Defaults to a day (24h). |
+| `COOKIE_SAME_SITE` | "Strict" | Allow you to specify the SameSite attribute of the Set-Cookie HTTP response header. Check details of SameSite attribute [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite). Its value could be "None", "Lax" or "Strict"  |
```
Comment (yanniszark) [inline README.md:99]: Can you make a small clarification:

name: `SESSION_SAME_SITE`

default: "Lax"

```
SameSite attribute of the session cookie. The [SameSite](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite) attribute acts as an extra protection against CSRF. Its value can be "None", "Lax" or "Strict".
```

File: server.go [hunk 2]
```diff
@@ -216,7 +217,15 @@ func (s *server) callback(w http.ResponseWriter, r *http.Request) {
...
-	session.Options.SameSite = http.SameSiteStrictMode
+	switch s.cookieSameSite {
+	case "None":
+		session.Options.SameSite = http.SameSiteNoneMode
+	case "Lax":
+		session.Options.SameSite = http.SameSiteLaxMode
+	default:
+		// Use Strict mode as default
+		session.Options.SameSite = http.SameSiteStrictMode
+	}
```
Comment (yanniszark) [inline server.go:228]: Can you make `Lax` the default?

File: settings.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ type config struct {
 	CABundlePath       string `split_words:"true" envconfig:"CA_BUNDLE"`
 	SessionStorePath   string `split_words:"true" default:"/var/lib/authservice/data.db"`
 	SessionMaxAge      int    `split_words:"true" default:"86400"`
+	CookieSameSite     string `split_words:"true" default:"Strict"`
```
Comment (yanniszark) [inline settings.go:50]: Same here, can you make `Lax` the default?
Also, can you use `SessionSameSite`, same as the setting?

File: main.go [hunk 4]
```diff
@@ -173,6 +173,7 @@ func main() {
 		caBundle:                caBundle,
 		authenticators:          []authenticator.Request{sessionAuthenticator, k8sAuthenticator},
 		authorizers:             []Authorizer{groupsAuthorizer},
+		cookieSameSite:           c.CookieSameSite,
```
Comment (yanniszark) [inline main.go:176]: `cookieSameSite` -> `sessionSameSite`

File: server.go [hunk 5]
```diff
@@ -53,6 +53,7 @@ type server struct {
 	idTokenOpts             jwtClaimOpts
 	upstreamHTTPHeaderOpts  httpHeaderOpts
 	caBundle                []byte
+	cookieSameSite          string
```
Comment (yanniszark) [inline server.go:56]: `cookieSameSite` -> `sessionSameSite`

**Review by yanniszark — CHANGES_REQUESTED** [review]
@yhwang thank you for your contribution!
In general, everything looks good. Just a few minor changes and we should be good to go!

**Comment by yanniszark** [comment]
In addition, can you expand a bit more in your commit message?
Similar to https://github.com/arrikto/oidc-authservice/pull/38:

> Using a strict SameSite policy for the session cookie is sometimes too
restrictive. For example, an application that is protected by the AuthService
may be acting as an OIDC/OAuth client to a third-party application. In that
case, the SameSite=strict policy prevents the Provider->Client auth code
redirect from working, as it originates from the Provider origin.

Make the SameSite policy of the session cookie configurable and default to
Lax, which allows cookies only from cross-site requests / redirects using the
GET method.

**Comment by yanniszark** [comment]
Finally, can you rebase to latest master?

File: README.md (hunk 1 above)
Comment (yhwang) [reply README.md]: done

File: server.go (hunk 2 above)
Comment (yhwang) [reply server.go]: sure, done

File: settings.go (hunk 3 above)
Comment (yhwang) [reply settings.go]: done

File: main.go (hunk 4 above)
Comment (yhwang) [reply main.go]: fixed

File: server.go (hunk 5 above)
Comment (yhwang) [reply server.go]: fixed

**Comment by yhwang** [comment]
I also rebased.

File: README.md [hunk 6]
```diff
@@ -96,6 +96,7 @@ Session store-related settings:
 | - | - | - |
 | `SESSION_STORE_PATH` | "/var/lib/authservice/data.db" | Path to local session store. Backed by BoltDB. |
 | `SESSION_MAX_AGE` | "86400" | Time in seconds after which sessions expire. Defaults to a day (24h). |
+| `SESSION_SAME_SITE` | "Lax" | Allow you to specify the SameSite attribute of the Set-Cookie HTTP response header. Check details of SameSite attribute [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite). Its value could be "None", "Lax" or "Strict"  |
```
Comment (yanniszark) [inline README.md:99]: Can you use the following text for description?

SameSite attribute of the session cookie. Check details of SameSite attribute [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite). Its value can be "None", "Lax" or "Strict".

**Review by yanniszark — CHANGES_REQUESTED** [review]
@yhwang thanks!
There are only 2 small things remaining.
1. Can you replace the setting description with the refactored one? I included it in a comment.
1. Can you please sign-off your commit? You can do so with `git commit -s --amend --no-edit`

File: README.md (hunk 6 above)
Comment (yhwang) [reply README.md]: done

**Comment by yanniszark** [comment]
Merged! Thanks a lot @yhwang

**Comment by yhwang** [comment]
thank YOU!

**Pull request closed without merging**